argus/action_manager/windows.py

```diff
@@ -115,32 +117,83 @@ def get_installation_script(self):
         self.download_resource("windows/installCBinit.ps1",
                                r"C:\installCBinit.ps1")
 
+    def _execute(self, cmd, count=util.RETRY_COUNT, delay=util.RETRY_DELAY,
+                 command_type=util.CMD):
+        """Execute until success and return only the standard output."""
+        stdout, _, _ = self._client.run_command_with_retry(
+            cmd, count=count, delay=delay, command_type=command_type)
+        return stdout
+
+    def check_cbinit_installation(self, python_dir):
+        """Check if Cloudbase-Init was installed successfully."""
+        LOG.info("Checking Cloudbase-Init installation.")
+
+        check_cmd = r'& "{}\python.exe" -c "import cloudbaseinit"'.format(
+            python_dir)
+
+        try:
+            self._client.run_remote_cmd(
+                cmd=check_cmd, command_type=util.POWERSHELL)
+        except exceptions.ArgusError as exc:
+            LOG.error("Cloudbase-Init installation failed: %s", exc)
+            return False
+
+        LOG.info("Cloudbase-Init was successfully installed!")
+        return True
+
+    def cbinit_cleanup(self, cbinit_dir):
+        """Clean up Cloudbase-Init."""
+        LOG.info("Cleaning up Cloudbase-Init ...")
+
+        self.rmdir(ntpath.dirname(cbinit_dir))
+
     def install_cbinit(self):
-        """Run the installation script for CloudbaseInit."""
-        LOG.debug("Installing Cloudbase-Init ...")
+        """Run the installation script for Cloudbase-Init."""
+        LOG.debug("Trying to install Cloudbase-Init ...")
 
         installer = "CloudbaseInitSetup_{build}_{arch}.msi".format(
             build=self._conf.argus.build,
             arch=self._conf.argus.arch
         )
         # TODO(cpopa): the service type is specific to each scenario,
         # find a way to pass it
-        LOG.info("Run the downloaded installation script "
-                 "using the installer %r.", installer)
 
-        parameters = '-installer {}'.format(installer)
-        try:
-            self.execute_powershell_resource_script(
-                resource_location='windows/installCBinit.ps1',
-                parameters=parameters)
-        except exceptions.ArgusError:
-            # This can happen for multiple reasons,
-            # but one of them is the fact that the installer
-            # can't be installed through WinRM on some OSes
-            # for whatever reason. In this case, we're falling back
-            # to use a scheduled task.
-            LOG.debug("Cannot install, deploying using a scheduled task.")
-            self._deploy_using_scheduled_task(installer)
+        for _ in range(util.RETRY_COUNT):
```

Comment:
This code block should be rewritten because it's hard to follow.

You can have a private method which will receive an action (the method of installing cloudbase-init), it will run that action and after that will check if everything it's okay (and if it's not it will clean up the environment).
If that method will return a bool it will be very easy to follow the workflow of the current method.

@stefan-caraiman @micumatei @ader1990 any other ideas ?